File: samples/AdventureWorksLT.VB/SalesOrderInfoDetail.vb
<Validator(SalesOrderInfoDetail._ValidateProduct)>
<InvisibleToDbDesigner>
Public Class SalesOrderInfoDetail
    Inherits SalesOrderDetail

    Shared Sub New()
        RegisterProjection(Function(x As SalesOrderInfoDetail) x.Product)
    End Sub

    Private m_Product As Product.Lookup
    Public Property Product As Product.Lookup
        Get
            Return m_Product
        End Get
        Private Set
            m_Product = Value
        End Set
    End Property

    Friend Const _ValidateProduct = NameOf(ValidateProduct)
    <_Validator>
    Private Function ValidateProduct(ByVal dataRow As DataRow) As DataValidationError
        If ProductID(dataRow) Is Nothing Then Return Nothing
        Dim productNumber = Product.ProductNumber
        Dim productName = Product.Name
        If String.IsNullOrEmpty(productNumber(dataRow)) Then Return New DataValidationError(String.Format(UserMessages.Validation_ValueIsRequired, productNumber.DisplayName), productNumber)
        If String.IsNullOrEmpty(productName(dataRow)) Then Return New DataValidationError(String.Format(UserMessages.Validation_ValueIsRequired, productName.DisplayName), productName)
        Return Nothing
    End Function
End Class

File: samples/CSharp/AdventureWorksLT/SalesOrderInfoDetail.cs
using DevZest.Data;
using DevZest.Data.Annotations;

namespace DevZest.Samples.AdventureWorksLT
{
    [Rule(nameof(ValidateProductNumber), SourceColumns = new string[] { nameof(ProductID), nameof(Product) + "." + nameof(SalesOrderInfoDetail.Product.ProductNumber) })]
    [Rule(nameof(ValidateProductName), SourceColumns = new string[] { nameof(ProductID), nameof(Product) + "." + nameof(SalesOrderInfoDetail.Product.Name) })]
    [InvisibleToDbDesigner]
    public class SalesOrderInfoDetail : SalesOrderDetail
    {
        static SalesOrderInfoDetail()
        {
            RegisterProjection((SalesOrderInfoDetail _) => _.Product);
        }

        public Product.Lookup Product { get; private set; }

        [_Rule]
        private DataValidationError ValidateProductNumber(DataRow dataRow)
        {
            if (ProductID[dataRow] == null)
                return null;
            var productNumber = Product.ProductNumber;

            if (string.IsNullOrEmpty(productNumber[dataRow]))
                return new DataValidationError(string.Format(UserMessages.Validation_ValueIsRequired, productNumber.DisplayName), productNumber);

            return null;
        }

        [_Rule]
        private DataValidationError ValidateProductName(DataRow dataRow)
        {
            if (ProductID[dataRow] == null)
                return null;

            var productName = Product.Name;
            if (string.IsNullOrEmpty(productName[dataRow]))
                return new DataValidationError(string.Format(UserMessages.Validation_ValueIsRequired, productName.DisplayName), productName);

            return null;
        }
    }
}
